rkt: add --hostname option

iaguis opened the pull request:
It allows the user to specify a host name for the pod. If not specified, it will be "rkt-$PODUUID".

This approach might not be OK if we want to use systemd-resolved, apparently it expects container's machine name to match the hostname.

cc @steveeJ @yifan-gu @sjpotter 

Fixes #2245

sjpotter commented:
@iaguis so would resolved break if a user changed the hostname within the container themselves?  and is that a bad thing?

iaguis commented:
I don't think I have enough context to answer that. Maybe @steveeJ can shed some light on it.

yifan-gu commented on stage1/init/init.go:
```diff
@@ -377,6 +369,18 @@ func getArgsEnv(p *stage1commontypes.Pod, flavor string, debug bool, n *networki
+	// According to systemd manual (https://www.freedesktop.org/software/systemd/man/hostname.html) :
+	// "The /etc/hostname file configures the name of the local system that is set
+	// during boot using the sethostname system call"
+	if hostname == "" {
```
This is not the common place. As the `case "kvm"` returns directly above.

iaguis replied:
Ooops

iaguis replied:
Updated, I don't know why this comment is still showing up on github...

jellonek replied:
Code was not changed, only moved, maybe this is a reason?

steveeJ commented:
@sjpotter 
> @iaguis so would resolved break if a user changed the hostname within the container themselves? and is that a bad thing?

This is not something we can easily answer in general, and therefore nothing we have to worry about immediately. You can read https://www.freedesktop.org/software/systemd/man/nss-mymachines.html# to get an idea what can influence the configuration. I just checked the `/etc/nsswitch.conf` inside CoreOS, which is also used within the default stage1, and either *mymachine* or *myhostname* is configured.

iaguis commented:
FWIW with `mymachines` in `/etc/nsswitch.conf`, resolution works fine with `rkt` and a different `--hostname`.

jellonek commented:
g2k
LGTM

iaguis commented:
If the resolved thing is not something we should worry about for now, I'm OK with merging this.

What do you thing, @steveeJ?

jellonek commented:
Is it possible that someone will add additional functional test for this before merge? It's not something what i need there strongly, but... (i know that i should ask about this @mstachowski when this was added to kvm flavor, but still...)

iaguis commented:
I can add one if we decide it's ok to merge :)

iaguis commented:
Added a functional test

steveeJ commented:
It will break fly :-(

steveeJ commented:
How about we introduce annotations for stage1 ACIs where supported commands (run, exec, gc) and their supported flags (--hostname, --dns, --port, ...) can be declared? stage0 could then check if it can pass all requested flags to the stage1 and either skip them or error out cleanly

jellonek commented:
Seems to me a bit too complicated, but reasonable. I have no strong feelings about this...

But... This will open a door for new features (and so incompatibility between flavors) supported by one flavor or group of them - this seems more scary and exciting ;)

iaguis commented:
> How about we introduce annotations for stage1 ACIs where supported commands (run, exec, gc) and their supported flags (--hostname, --dns, --port, ...) can be declared? stage0 could then check if it can pass all requested flags to the stage1 and either skip them or error out cleanly. Behavior on incompatibility should default to error but also be settable via a stage0 flag.

This is a pretty big change, can we just ignore the flag in fly for now so we can get this and discuss that in a new issue?

alban commented on Documentation/subcommands/run-prepared.md:
```diff
@@ -48,6 +48,7 @@ c9fad0e6    etcd    coreos.com/etcd prepared
 | `--dns` |  `` | IP Address | Name server to write in `/etc/resolv.conf`. It can be specified several times |
 | `--dns-opt` |  `` | Option as described in the options section in resolv.conf(5) | DNS option to write in `/etc/resolv.conf`. It can be specified several times |
 | `--dns-search` |  `` | Domain name | DNS search domain to write in `/etc/resolv.conf`. It can be specified several times |
+| `--hostname` |  `` | A host name | Pod's hostname. If empty, it will be the "rkt-$PODUUID" |
```
s/the//

alban commented on Documentation/subcommands/run.md:
```diff
@@ -314,6 +314,7 @@ If you don't use systemd, you can use [daemon](http://www.libslack.org/daemon/)
 | `--dns-opt` |  `` | Option as described in the options section in resolv.conf(5) | DNS option to write in `/etc/resolv.conf`. It can be specified several times |
 | `--dns-search` |  `` | Domain name | DNS search domain to write in `/etc/resolv.conf`. It can be specified several times |
 | `--exec` |  `` | A path | Override the exec command for the preceding image |
+| `--hostname` |  `` | A host name | Pod's hostname. If empty, it will be the "rkt-$PODUUID" |
```
ditto

iaguis commented:
> This is a pretty big change, can we just ignore the flag in fly for now (like we do with the rest) so we can get this and discuss that in a new issue?

I added a commit doing that, addressed the review and added `--hostname` to the stage1 implementors guide.

alban commented:
The `--hostname` is not in the documentation:
https://github.com/coreos/rkt/blob/master/Documentation/devel/stage1-implementors-guide.md#arguments

I am not sure about adding a new parameter without introducing versioning in annotations as explained here:
https://github.com/coreos/rkt/pull/2173#issuecomment-182841553

steveeJ commented:
>> How about we introduce annotations for stage1 ACIs where supported commands (run, exec, gc) and their supported flags (--hostname, --dns, --port, ...) can be declared? stage0 could then check if it can pass all requested flags to the stage1 and either skip them or error out cleanly. Behavior on incompatibility should default to error but also be settable via a stage0 flag.
> 
> This is a pretty big change, can we just ignore the flag in fly for now (like we do with the rest) so we can get this and discuss that in a new issue?

Could you break it apart to identify the most difficult/time-consuming steps that may be postponed?

jonboulle commented:
> This is a pretty big change, can we just ignore the flag in fly for now (like we do with the rest) so we can get this and discuss that in a new issue?

Let's do this. There are many outstanding issues with the fly/stage1 thing that need to be addressed, but adding another ignored flag is only making the situation marginally worse.

yifan-gu commented:
I think this also affects how we get logs from journactl. `journalctl -m _HOSTNAME=rkt-uuid` doesn't seem to work now in this example: https://github.com/coreos/rkt/blob/master/Documentation/using-rkt-with-systemd.md#systemd-run

iaguis commented:
You can do `journalctl -m _MACHINE_ID=${UUID_WITHOUT_DASHES}`.

krnowak commented on Documentation/devel/stage1-implementors-guide.md:
```diff
@@ -60,6 +60,7 @@ Stage 1 must write the host PIDs of the pod's process #1 and that process's pare
 * `--net[=$NET1,$NET2,...]` to configure the creation of a contained network.
   See the [rkt networking documentation](../networking.md) for details.
 * `--mds-token=$TOKEN` passes the auth token to the apps via `AC_METADATA_URL` env var
+* `--hostname=$HOSTNAME` configures the host name of the pod. If empty, it will be "rkt-$PODUUID". Only available from ABI version 2.
```
Probably would be easier to read if there was a separate section like "flags added in interface version 2".

krnowak commented on stage0/run.go:
```diff
@@ -491,6 +492,15 @@ func Run(cfg RunConfig, dir string, dataDir string) {
+		log.FatalE("error determining stage1 interface version", err)
+	}
+
+	if cfg.Hostname != "" && s1v > 1 {
```
I'd print a warning if we specified `--hostname`, but stage1 interface version is < 2.

sjpotter commented:
yea, we figured that out, but strangely, it doesn't work on my ubuntu box.
 ( _HOSTNAME or _MACHINE_ID)

On Thu, Mar 17, 2016 at 9:01 AM, Iago López Galeiras <
notifications@github.com> wrote:

> You can do journalctl -m _MACHINE_ID=${UUID_WITHOUT_DASHES}.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/coreos/rkt/pull/2251#issuecomment-197948383>
>

iaguis commented:
Updated.

alban commented on stage0/run.go:
```diff
@@ -491,6 +492,19 @@ func Run(cfg RunConfig, dir string, dataDir string) {
+	}
+
+	if cfg.Hostname != "" {
+		if s1v > 1 {
```
I wonder about defining a constant in interface.go, and then:
```
if s1v > interfaceVersionHostname {
```

krnowak replied:
Dunno, it might be troublesome later. interfaceVersionThisThatAndOneMoreThingIfIRemeberedEverything.

iaguis replied:
That's how German works, right?

alban replied:
actually, instead of a constant, define a function in interface.go:
```
if interfaceVersionSupportsHostname(s1v) {
}
```

krnowak replied:
Or maybe make s1v an object instead of an integer, so you could add functions to it like `func (s1v Stage1InterfaceVersion) SupportsHostname() bool {}`

That would be an idea for a separate PR. ;P

iaguis replied:
I'll think about it... :P

krnowak commented:
LFAD, but maybe you'll want another stamp.

alban commented:
LGTM if green
